src/service/position-observer/position-observer-impl.js

```diff
@@ -0,0 +1,236 @@
+/**
+ * Copyright 2017 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use baseInstance file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {registerServiceBuilderForDoc} from '../../service';
+import {Services} from '../../services';
+import {dev} from '../../log';
+import {layoutRectLtwh} from '../../layout-rect';
+import {debounce} from '../../utils/rate-limit';
+import {PositionObserverEntry} from './position-observer-entry';
+import {
+  PositionObserverFidelity,
+  LOW_FIDELITY_FRAME_COUNT,
+} from './position-observer-fidelity';
+/** @const @private */
+const TAG = 'POSITION_OBSERVER';
+
+/** @const @private */
+const SCROLL_TIMEOUT = 500;
+
+
+export class PositionObserver {
+  /**
+   * @param {!../ampdoc-impl.AmpDoc} ampdoc
+   */
+  constructor(ampdoc) {
+    /** @private {!Window} */
+    this.win_ = ampdoc.win;
+
+    /** @private {!Array<!PositionObserverEntry>} */
+    this.entries_ = [];
+
+    /** @private {!../vsync-impl.Vsync} */
+    this.vsync_ = Services.vsyncFor(this.win_);
+
+    /** @private {!../viewport-impl.Viewport} */
+    this.viewport_ = Services.viewportForDoc(ampdoc);
+
+    /** @private {Array<function()>} */
+    this.unlisteners_ = [];
+
+    /** @private {boolean} */
+    this.inScroll_ = false;
+
+    /** @private {boolean} */
+    this.measure_ = false;
+
+    /** @private {function()} */
+    this.boundStopScroll_ = debounce(this.win_, () => {
+      this.inScroll_ = false;
+    }, SCROLL_TIMEOUT);
+  }
+
+  /**
+   * @param {!Element} element
+   * @param {PositionObserverFidelity} fidelity
+   * @param {function(?./position-observer-entry.PositionInViewportEntryDef)} handler
+   */
+  observe(element, fidelity, handler) {
+    const entry = new PositionObserverEntry(element, fidelity, handler);
+
+    this.entries_.push(entry);
+
+    if (this.entries_.length == 1) {
```

Comment:
`this.entries_.length == 1` was confusing to me initially had to think to see why. Can we track through another variable? like `callbackStarted_`?

Reply:
added